src/dotty/tools/dotc/transform/ClassTags.scala

```diff
@@ -0,0 +1,57 @@
+package dotty.tools.dotc
+package transform
+
+import core._
+import TreeTransforms._
+import Contexts.Context
+import Flags._
+import SymUtils._
+import Symbols._
+import SymDenotations._
+import Types._
+import Decorators._
+import DenotTransformers._
+import StdNames._
+import NameOps._
+import ast.Trees._
+import dotty.tools.dotc.ast.tpd
+import dotty.tools.dotc.core.Constants.Constant
+import util.Positions._
+import Names._
+import collection.mutable
+
+/** This phase replaces calls to DottyPredef.classTag by code that synthesizes appropriate ClassTag
+  */
+class ClassTags extends MiniPhaseTransform with IdentityDenotTransformer { thisTransform =>
+  import ast.tpd._
+
+  object names {
+    val classTag = "classTag".toTermName
+    val typeTag  = "typeTag".toTermName
+  }
+
+  private var classTagCache: Symbol = null
+  private var typeTagCache: Symbol  = null
+  private var scala2ClassTagModule: Symbol = null
+
+
+  override def prepareForUnit(tree: tpd.Tree)(implicit ctx: Context): TreeTransform = {
+    val predefClass = defn.DottyPredefModule.moduleClass.asClass
+    classTagCache = ctx.requiredMethod(predefClass,  names.classTag)
+    typeTagCache = ctx.requiredMethod(predefClass,  names.typeTag)
+    scala2ClassTagModule = ctx.requiredModule("scala.reflect.ClassTag")
+    this
+  }
+
+  override def phaseName: String = "classTags"
+
+  override def transformTypeApply(tree: tpd.TypeApply)(implicit ctx: Context, info: TransformerInfo): tpd.Tree =
+    if (tree.fun.symbol eq classTagCache) {
+      val tp = tree.args.head.tpe
+      val claz = tp.classSymbol
```

Comment:
On Thu, May 14, 2015 at 9:07 AM, Dmitry Petrashko <notifications@github.com>
wrote:

> In src/dotty/tools/dotc/transform/ClassTags.scala
> <https://github.com/lampepfl/dotty/pull/563#discussion_r30301852>:
>
> > +
> > +
> > +  override def prepareForUnit(tree: tpd.Tree)(implicit ctx: Context): TreeTransform = {
> > +    val predefClass = defn.DottyPredefModule.moduleClass.asClass
> > +    classTagCache = ctx.requiredMethod(predefClass,  names.classTag)
> > +    typeTagCache = ctx.requiredMethod(predefClass,  names.typeTag)
> > +    scala2ClassTagModule = ctx.requiredModule("scala.reflect.ClassTag")
> > +    this
> > +  }
> > +
> > +  override def phaseName: String = "classTags"
> > +
> > +  override def transformTypeApply(tree: tpd.TypeApply)(implicit ctx: Context, info: TransformerInfo): tpd.Tree =
> > +    if (tree.fun.symbol eq classTagCache) {
> > +      val tp = tree.args.head.tpe
> > +      val claz = tp.classSymbol
>
> Thanks for Array example, It's indeed not handled correctly.
> classSymbol for OrTypes is bugged, see comment
> https://github.com/lampepfl/dotty/blob/master/src/dotty/tools/dotc/core/Types.scala#L308.
> It should not return NoSymbol.
> Though this code will not work for some AndTypes.
>
> It seems that classSymbol.orElse(classSymbols.head) should do it
> correctly.
>
> That will return a symbol. but how do you know it's the right one? We are
missing a spec here. I think the most robust would be to take the
classSymbol of the erased type. That has a simple spec and it is likely to
be the one we will want in the end.


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/pull/563/files#r30301852>.
>



-- 
Martin Odersky
EPFL